race-buffer/src/lib.rs

```diff
@@ -268,62 +79,97 @@ pub mod tests {
             OrderedEntry(i + Self::SUFFIX_TAG)
         }
 
+        pub(crate) fn is_prefix_unchecked(&self) -> bool {
+            self.0 > Self::PREFIX_TAG
+        }
+
         pub(crate) fn is_suffix(&self) -> bool {
             self.0 >= Self::SUFFIX_TAG && self.0 < Self::PREFIX_TAG
         }
 
         // Invariant: Check if entries all have correct index
-        fn entries_correct_index(rbuf: &[OrderedEntry]) -> bool {
+        fn entries_correct_index(rbuf: &[Option<OrderedEntry>]) -> bool {
             for i in 0..rbuf.len() {
-                if rbuf[i].to_raw() != i as u32 + 1 && rbuf[i] != Self::NIL_VAL {
-                    return false;
+                if let Some(entry) = rbuf[i] {
+                    if entry.to_index() != i as u32 + 1 {
+                        return false;
+                    }
                 }
             }
             return true;
         }
 
         // Invariant: Check if entries all have correct index
-        fn double_entries_consistent(rbuf: &[OrderedEntry]) -> bool {
-            for i in 0..(rbuf.len() - 1) {
-                let cur = &rbuf[i];
-                let next = &rbuf[i + 1];
-                if (cur.is_prefix() && !next.is_suffix()) || (!cur.is_prefix() && next.is_suffix())
-                {
+        fn double_entries_consistent(rbuf: &[Option<OrderedEntry>]) -> bool {
+            if let Some(first_entry) = rbuf[0] {
+                if first_entry.is_suffix() {
                     return false;
                 }
             }
+            if let Some(last_entry) = rbuf[rbuf.len() - 1] {
```

Comment:
The Rust idiom for this is [last()](https://doc.rust-lang.org/std/primitive.slice.html#method.last).